### PR TITLE
[SPARK-27641][CORE] Fix MetricsSystem to remove metrics of the specific source by exactly match

### PR DESCRIPTION
**Pull request by chummyhe89**
## What changes were proposed in this pull request?

* Only remove the registered metrics  of the specific source when perform `MetricsSystem.removeSource`

## How was this patch tested?
* add new unit tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 1]
```diff
@@ -155,6 +156,12 @@ private[spark] class MetricsSystem private (
     } else { defaultName }
   }
 
+  private[spark] def getRegisteredNames(source: Source): mutable.Set[String] = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: Does not need to be mutable, or a Set. Just Seq?

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 2]
```diff
@@ -169,9 +176,12 @@ private[spark] class MetricsSystem private (
   }
 
   def removeSource(source: Source): Unit = {
-    sources -= source
-    val regName = buildRegistryName(source)
-    registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
+    val index = sources.indexOf(source)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: Do you just want to check if `sources.contains` the source?

**Review by srowen — COMMENTED** [review]
Probably OK; is there any reason one would want to keep the metrics around after removing the source?
Would this clear all metrics on shutdown?

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala (hunk 1 above)
Comment (chummyhe89) [reply core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: you are right. thanks for your advice.

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala (hunk 2 above)
Comment (chummyhe89) [reply core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: `sources` is an ArrayBuffer.  if  `sources` really contains the checking one,the index can avoid traversing array twice.

**Comment by jiangxb1987** [comment]
What's the behavior change you are proposing? Your new test case passes under current code base.

**Pull request closed without merging**